tpu-client-next/src/node_address_service/leader_tpu_cache_service.rs

```diff
@@ -0,0 +1,520 @@
+    }
+}
+
+#[derive(PartialEq, Debug)]
+struct SlotLeaders {
+    first_slot: Slot,
+    leaders: Vec<Pubkey>,
+}
+
+impl SlotLeaders {
+    async fn new(
+        rpc_client: &impl ClusterInfoProvider,
+        estimated_current_slot: Slot,
+        slots_in_epoch: Slot,
+    ) -> Result<Self, Error> {
+        Ok(Self {
+            first_slot: estimated_current_slot,
+            leaders: rpc_client
+                .slot_leaders(estimated_current_slot, slots_in_epoch)
+                .await?,
+        })
+    }
+
+    fn last_slot(&self) -> Slot {
+        self.first_slot + self.leaders.len().saturating_sub(1) as u64
+    }
+
+    fn slot_leader(&self, slot: Slot) -> Option<&Pubkey> {
+        slot.checked_sub(self.first_slot)
+            .and_then(|index| self.leaders.get(index as usize))
+    }
+
+    fn is_last_slot_in_window(&self, slot: Slot) -> Option<bool> {
+        slot.checked_sub(self.first_slot).and_then(|index| {
+            let index = index as usize;
+            if index + 1 < self.leaders.len() {
+                Some(self.leaders[index] != self.leaders[index + 1])
+            } else {
+                None
+            }
+        })
+    }
+}
+
+#[derive(PartialEq, Debug)]
+struct EpochInfo {
+    slots_in_epoch: Slot,
+    last_slot_in_epoch: Slot,
+}
+
+impl EpochInfo {
+    async fn new(
+        rpc_client: &impl ClusterInfoProvider,
+        estimated_current_slot: Slot,
+    ) -> Result<Self, Error> {
+        let (slots_in_epoch, last_slot_in_epoch) =
+            rpc_client.epoch_info(estimated_current_slot).await?;
+        Ok(Self {
+            slots_in_epoch,
+            last_slot_in_epoch,
+        })
+    }
+}
+
+#[async_trait]
+impl ClusterInfoProvider for RpcClient {
+    async fn leader_tpu_map(&self) -> Result<HashMap<Pubkey, SocketAddr>, Error> {
+        let cluster_nodes = self.get_cluster_nodes().await.map_err(Error::RpcError)?;
+        Ok(extract_cluster_tpu_sockets(cluster_nodes))
+    }
+
+    async fn epoch_info(&self, estimated_current_slot: Slot) -> Result<(Slot, Slot), Error> {
+        match self.get_epoch_schedule().await {
+            Ok(epoch_schedule) => {
+                let epoch = epoch_schedule.get_epoch(estimated_current_slot);
+                let slots_in_epoch = epoch_schedule.get_slots_in_epoch(epoch);
+                let last_slot_in_epoch = epoch_schedule.get_last_slot_in_epoch(epoch);
+                debug!(
+                    "Updated slots in epoch: {slots_in_epoch}, last slot in epoch: \
+                     {last_slot_in_epoch}",
+                );
+                Ok((slots_in_epoch, last_slot_in_epoch))
+            }
+            Err(err) => Err(Error::RpcError(err)),
+        }
+    }
+
+    async fn slot_leaders(
+        &self,
+        estimated_current_slot: Slot,
+        slots_in_epoch: Slot,
+    ) -> Result<Vec<Pubkey>, Error> {
+        let slot_leaders = self
+            .get_slot_leaders(estimated_current_slot, fanout(slots_in_epoch))
```

Comment:
ARAFIK get_slot_leaders can look at max 5000 slots ahead; so capping to `slots_in_epoch` might not help

{
  "jsonrpc": "2.0",
  "error": {
    "code": -32602,
    "message": "Invalid limit; max 5000"
  },
  "id": 1
}

>